library/src/main/java/com/datatorrent/lib/join/AbstractJoinOperator.java

```diff
@@ -0,0 +1,343 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.lib.join;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+
+import org.apache.hadoop.classification.InterfaceStability.Evolving;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.Operator;
+import com.datatorrent.api.annotation.InputPortFieldAnnotation;
+import com.datatorrent.common.util.BaseOperator;
+
+/**
+ * <p>
+ * This is the base implementation of join operator. Operator receives tuples from multiple streams,
+ * applies the join operation based on constraint and emit the joined value.
+ * Subclasses should provide implementation to createOutputTuple,copyValue, getKeyValue, getTime methods.
+ *
+ * <b>Properties:</b><br>
+ * <b>expiryTime</b>: Expiry time for stored tuples<br>
+ * <b>includeFieldStr</b>: List of comma separated fields to be added to the output tuple.
+ *                         Ex: Field1,Field2;Field3,Field4<br>
+ * <b>keyFields</b>: List of comma separated key field for both the streams. Ex: Field1,Field2<br>
+ * <b>timeFields</b>: List of comma separated time field for both the streams. Ex: Field1,Field2<br>
+ * <b>bucketSpanInMillis</b>: Span of each bucket in milliseconds.<br>
+ * <b>strategy</b>: Type of join operation. Default type is inner join<br>
+ * <br>
+ * </p>
+ *
+ * @displayName Abstract Join Operator
+ * @tags join
+ */
+@Evolving
+public abstract class AbstractJoinOperator<T> extends BaseOperator implements Operator.CheckpointListener
+{
+  public final transient DefaultOutputPort<List<T>> outputPort = new DefaultOutputPort<>();
+  protected String[] timeFields;
+
+  protected String[][] includeFields;
+  // Fields to compare from both the streams
+  protected String[] keys;
+  // Strategy of Join operation, by default the option is inner join
+  protected JoinStrategy strategy = JoinStrategy.INNER_JOIN;
+  // This represents whether the processing tuple is from left port or not
+  protected Boolean isLeft;
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<T> input1 = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      isLeft = true;
+      processTuple(tuple);
+    }
+  };
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<T> input2 = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      isLeft = false;
+      processTuple(tuple);
+    }
+  };
+  // Stores for each of the input port
+  private BackupStore[] store = (BackupStore[])Array.newInstance(BackupStore.class, 2);
+  private String includeFieldStr;
+  private String keyFieldStr;
+  private String timeFieldStr;
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    if (store[0] == null) {
```

Comment:
@chaithu14 if store are annotated as @NotNull then this can be validated before launch. I think you made store as a 2 element array this is why you have to check it at run time.
There are validation constraints that you can explore to make this validation happen before launch  for an array but I will suggest to have the 2 stores as different fields.